velox/common/compression/Compression.h

```diff
@@ -45,6 +48,201 @@ CompressionKind stringToCompressionKind(const std::string& kind);
+
+  /// Return the default compression level.
+  /// If the codec doesn't support compression level,
+  /// `kUseDefaultCompressionLevel` will be returned.
+  virtual int32_t defaultCompressionLevel() const = 0;
+
+  /// Performs one-shot compression.
+  /// `outputLength` must first have been computed using maxCompressedLength().
+  /// The actual compressed length will be written to actualOutputLength.
+  /// Note: One-shot compression is not always compatible with streaming
+  /// decompression. Depending on the codec (e.g. LZ4), different formats may
+  /// be used.
+  virtual Expected<uint64_t> compress(
+      const uint8_t* input,
+      uint64_t inputLength,
+      uint8_t* output,
+      uint64_t outputLength) = 0;
+
+  /// One-shot decompression function.
+  /// `outputLength` must be correct and therefore be obtained in advance.
+  /// The actual decompressed length is returned.
+  /// Note: One-shot decompression is not always compatible with streaming
+  /// compression. Depending on the codec (e.g. LZ4), different formats may
+  /// be used.
+  virtual Expected<uint64_t> decompress(
+      const uint8_t* input,
+      uint64_t inputLength,
+      uint8_t* output,
+      uint64_t outputLength) = 0;
+
+  /// Performs one-shot compression.
+  /// This function compresses data and writes the output up to the specified
+  /// outputLength. If outputLength is too small to hold all the compressed
+  /// data, the function doesn't fail. Instead, it returns the number of bytes
+  /// actually written to the output buffer. Any remaining data that couldn't
+  /// be written in this call will be written in subsequent calls to this
+  /// function. This is useful when fixed-length compression blocks are required
+  /// by the caller.
+  /// Note: Only Gzip and Zstd codec supports this function.
+  virtual Expected<uint64_t> compressFixedLength(
+      const uint8_t* input,
+      uint64_t inputLength,
+      uint8_t* output,
+      uint64_t outputLength);
+
+  // Maximum compressed length of given input length.
+  virtual uint64_t maxCompressedLength(uint64_t inputLength) = 0;
+
+  /// Retrieves the actual uncompressed length of data using the specified
+  /// compression library.
+  /// Note: This functionality is not universally supported by all compression
+  /// libraries. If not supported, `std::nullopt` will be returned.
+  virtual std::optional<uint64_t> getUncompressedLength(
+      const uint8_t* input,
+      uint64_t inputLength) const;
+
+  // Create a streaming compressor instance.
+  virtual Expected<std::shared_ptr<StreamingCompressor>>
+  makeStreamingCompressor();
+
+  // Create a streaming compressor instance.
+  virtual Expected<std::shared_ptr<StreamingDecompressor>>
+  makeStreamingDecompressor();
+
+  // This Codec's compression type.
+  virtual CompressionKind compressionKind() const = 0;
+
+  // This Codec's compression level, if applicable.
+  virtual int32_t compressionLevel() const;
```

Comment:
Not virtual? It is set by static create function.

Reply:
For codecs supporting the compression level, it returns the compression level specified in CodecOptions. While some codecs may not support setting compression level, such as snappy. In this case kUseDefaultCompressionLevel is returned. https://github.com/marin-ma/velox-oap/commit/2dc70d568786a62ea3396b64c63e6e3f07e40eaa#diff-5da49de45811211ca959d81031d79e7fb89eb65848c4a7647d9dd7ded1ee4403R62

> Is it also possible the compressionKind is supported in streaming compressor but compression level is mot satisfied?

No.